### PR TITLE
Resume Signal Form

### PR DESCRIPTION
**Pull request by james-union**
https://github.com/flyteorg/flyteconsole/issues/587

# TL;DR
- SetSignal is still waiting to be deployed on the backend side

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
fixes https://github.com/flyteorg/flyteconsole/issues/587

### CODE REVIEW

File: packages/zapp/console/src/components/Launch/LaunchForm/constants.ts [hunk 1]
```diff
@@ -4,13 +4,17 @@ import { BlobValue, InputType, AuthRoleTypes, AuthRoleMeta } from './types';
...
   submit: 'Launch',
+  resume: 'Resume',
   taskVersion: 'Task Version',
   title: 'Create New Execution',
+  resumeTitle: 'Resume Paused Execution',
   workflowVersion: 'Workflow Version',
   launchPlan: 'Launch Plan',
   interruptible: 'Interruptible',
+  viewNodeInputs: 'View node inputs',
 };
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/constants.ts:18]: let's move all of the formStrings and lines 77-90 (except line 83 `workflowNoInputsString`, as it's being imported outside of the LaunchForm too) in a localization strings file for this folder

File: packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import * as React from 'react';
+import { createInputValueCache, InputValueCacheContext } from './inputValueCache';
+import { ResumeSignalForm } from './ResumeSignalForm';
+import { ResumeSignalFormProps } from './types';
+
+/** Renders the form for initiating a Launch request based on a Workflow or Task */
+export const ResumeForm: React.FC<ResumeSignalFormProps> = (props) => {
+  const [inputValueCache] = React.useState(createInputValueCache());
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx:8]: update hook import

File: packages/zapp/console/src/components/Executions/contextProvider/NodeExecutionDetails/createExecutionArray.tsx [hunk 3]
```diff
@@ -3,7 +3,7 @@ import { getTaskDisplayType } from 'components/Executions/utils';
 import { NodeExecutionDetails, NodeExecutionDisplayType } from 'components/Executions/types';
 import { Workflow } from 'models/Workflow/types';
 import { Identifier } from 'models/Common/types';
-import { CompiledTask } from 'models/Task/types';
+import { CompiledTask, TaskTemplate } from 'models/Task/types';
```
Comment (olga-union) [inline packages/zapp/console/src/components/Executions/contextProvider/NodeExecutionDetails/createExecutionArray.tsx:6]: remove unused type import

**Review by olga-union — COMMENTED** [review]
_Don't forget to include `flyteidl` update in this PR_ - ignore for now, looks like v1.1.22 doesn't have gate nodes

File: packages/zapp/console/src/components/Launch/LaunchForm/types.ts [hunk 4]
```diff
@@ -5,8 +5,9 @@ import {
   LiteralType,
   NamedEntityIdentifier,
 } from 'models/Common/types';
-import { WorkflowExecutionIdentifier } from 'models/Execution/types';
+import { NodeExecutionIdentifier, WorkflowExecutionIdentifier } from 'models/Execution/types';
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/types.ts:8]: `NodeExecutionIdentifier` is unused

File: packages/zapp/console/src/components/Launch/LaunchForm/useResumeFormState.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+export function useResumeFormState({ compiledNode }: ResumeFormProps): ResumeFormState {
+  const apiContext = useAPIContext();
+  const formInputsRef = useRef<LaunchFormInputsRef>(null);
+
+  const services = useMemo(
+    () => getServices(apiContext, formInputsRef),
+    [apiContext, formInputsRef],
+  );
+
+  const [state, sendEvent, service] = useMachine<
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/useResumeFormState.ts:99]: `sendEvent` is unused

File: packages/zapp/console/src/components/Launch/LaunchForm/ResumeFormHeader.tsx [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+export const ResumeFormHeader: React.FC<{ title?: string }> = ({ title = '' }) => {
+  const styles = useStyles();
+  return (
+    <DialogTitle disableTypography={true} className={styles.header}>
+      <div className={styles.inputLabel}>{formStrings.resumeTitle}</div>
+      <Typography variant="h6">{title}</Typography>
+    </DialogTitle>
+  );
+};
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/ResumeFormHeader.tsx:14]: This is 99% duplicate of `LaunchFormHeader`, we need to reuse the existing component and just pass down appropriate `formTitle` prop based on the place it's being rendered (formStrings.resumeTitle or formStrings.title)

File: packages/zapp/console/src/components/Launch/LaunchForm/ResumeFormActions.tsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
+import { Button, DialogActions, FormHelperText } from '@material-ui/core';
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/ResumeFormActions.tsx:1]: same here - 99% copy-paste, reuse `LaunchFormAction` and pass an extra prop with button title

File: packages/zapp/console/src/components/Launch/LaunchForm/ResumeFormDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import { Dialog } from '@material-ui/core';
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/ResumeFormDialog.tsx:1]: This one is ~90% similarity, so no need for a separate component too.
Next props would be optional: `id`, `compiledNode`, `nodeId`, so inside the `Dialog` you'll have a ternary `id ? <LaunchForm /> : {compiledNode && nodeId && <ResumeForm />}` or something like that

File: packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx (hunk 2 above)
Comment (olga-union) [reply packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx:8]: Ok, I don't think you use `inputValueCache` anywhere in the resume form, so we don't need this wrapper - `<ResumeSignalForm />` can be used directly in its place

Comment (olga-union) [reply packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx:8]: even better would be to rename `ResumeSignalForm` as `ResumeForm` and delete this component

File: packages/zapp/console/src/components/Launch/LaunchForm/useResumeFormState.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+  ResumeFormState,
+  TaskInitialLaunchParameters,
+} from './types';
+import { getUnsupportedRequiredInputs } from './utils';
+
+export interface ResumeFormProps extends BaseLaunchFormProps {
+  compiledNode: CompiledNode;
+  initialParameters?: TaskInitialLaunchParameters;
+  nodeId: string;
+}
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/useResumeFormState.ts:29]: unused it this file, and has the same definition in `ResumeForm` component

File: packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+import { useState } from 'react';
+import { createInputValueCache, InputValueCacheContext } from './inputValueCache';
+import { ResumeSignalForm } from './ResumeSignalForm';
+import { BaseLaunchFormProps, TaskInitialLaunchParameters } from './types';
+
+export interface ResumeFormProps extends BaseLaunchFormProps {
+  compiledNode: CompiledNode;
+  initialParameters?: TaskInitialLaunchParameters;
+  nodeId: string;
+}
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx:12]: remove `export` - it's a local interface

File: packages/zapp/console/src/components/Launch/LaunchForm/ResumeSignalForm.tsx [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  nodeId,
+  onClose,
+}) => {
+  const { formInputsRef, state, service } = useResumeFormState({ compiledNode, nodeId, onClose });
+  const nodeExecutionsById = useContext(NodeExecutionsByIdContext);
+  const [nodeExecution, setNodeExecution] = useState<NodeExecution>(nodeExecutionsById[nodeId]);
+  const styles = useStyles();
+  const baseState = state as BaseInterpretedLaunchState;
+  const baseService = service as BaseLaunchService;
+  const [isError, setIsError] = React.useState<boolean>(false);
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/ResumeSignalForm.tsx:43]: React. leftover

File: packages/zapp/console/src/components/Launch/LaunchForm/LaunchFormDialog.tsx [hunk 12]
```diff
@@ -24,7 +28,7 @@ function getLaunchProps(id: ResourceIdentifier) {
 }
 
 export const LaunchFormDialog = (props: LaunchFormDialogProps): JSX.Element => {
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/LaunchFormDialog.tsx:30]:
```suggestion
export const LaunchFormDialog = ({ id, initialParameters, showLaunchForm, setShowLaunchForm, compiledNode, nodeId }: LaunchFormDialogProps): JSX.Element => {
```

File: packages/zapp/console/src/components/Launch/LaunchForm/LaunchFormDialog.tsx [hunk 13]
```diff
@@ -24,7 +28,7 @@ function getLaunchProps(id: ResourceIdentifier) {
...
-  const { id, initialParameters, showLaunchForm, setShowLaunchForm } = props;
+  const { id, initialParameters, showLaunchForm, setShowLaunchForm, compiledNode, nodeId } = props;
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/LaunchFormDialog.tsx:31]:
```suggestion
```

File: packages/zapp/console/src/components/Executions/Tables/NodeExecutionActions.tsx [hunk 14]
```diff
@@ -122,7 +105,26 @@ export const NodeExecutionActions = ({ execution }: NodeExecutionActionsProps):
...
+        <IconButton onClick={rerunIconOnClick}>
+          <RerunIcon />
+        </IconButton>
+      </Tooltip>
+      <LaunchFormDialog
+        id={id as ResourceIdentifier}
+        initialParameters={initialParameters}
+        showLaunchForm={showLaunchForm}
+        setShowLaunchForm={setShowLaunchForm}
+      />
```
Comment (olga-union) [inline packages/zapp/console/src/components/Executions/Tables/NodeExecutionActions.tsx:118]: I think this should be wrapped in id and initialParameters presence check
```suggestion
      {id && initialParameters && (<Tooltip title={t('rerunTooltip')}>
        <IconButton onClick={rerunIconOnClick}>
          <RerunIcon />
        </IconButton>
      </Tooltip>
      <LaunchFormDialog
        id={id as ResourceIdentifier}
        initialParameters={initialParameters}
        showLaunchForm={showLaunchForm}
        setShowLaunchForm={setShowLaunchForm}
      />
      )}
```

File: packages/zapp/console/src/components/Launch/LaunchForm/LaunchInterruptibleInput.tsx [hunk 15]
```diff
@@ -74,14 +74,14 @@ export const LaunchInterruptibleInputImpl: React.ForwardRefRenderFunction<
   const getInterruptibleLabel = () => {
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/LaunchInterruptibleInput.tsx:74]: Can you please add TODO above this function to cover all text variants in localization conditional?

File: packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+import { ResumeSignalForm } from './ResumeSignalForm';
+import { BaseLaunchFormProps, TaskInitialLaunchParameters } from './types';
+
+interface ResumeFormProps extends BaseLaunchFormProps {
+  compiledNode: CompiledNode;
+  initialParameters?: TaskInitialLaunchParameters;
+  nodeId: string;
+}
+
+/** Renders the form for initiating a Launch request based on a Workflow or Task */
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx:14]: This description doesn't match the purpose of the component

File: packages/zapp/console/src/components/Launch/LaunchForm/ResumeSignalForm.tsx [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+import launchFormStrings from './strings';
+import { LaunchFormActions } from './LaunchFormActions';
+
+export interface ResumeSignalFormProps extends BaseLaunchFormProps {
+  compiledNode: CompiledNode;
+  initialParameters?: TaskInitialLaunchParameters;
+  nodeId: string;
+}
+
+/** Renders the form for initiating a Launch request based on a Task */
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/ResumeSignalForm.tsx:31]: this one is related to original LaunchForm too 😄

File: packages/zapp/console/src/components/Launch/LaunchForm/ResumeSignalForm.tsx [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+  const [nodeExecution, setNodeExecution] = useState<NodeExecution>(nodeExecutionsById[nodeId]);
+  const styles = useStyles();
+  const baseState = state as BaseInterpretedLaunchState;
+  const baseService = service as BaseLaunchService;
+  const [isError, setIsError] = useState<boolean>(false);
+  const executionData = useNodeExecutionData(nodeExecution.id);
+
+  // Any time the inputs change (even if it's just re-ordering), we must
+  // change the form key so that the inputs component will re-mount.
+  const formKey = React.useMemo<string>(() => {
```
Comment (olga-union) [inline packages/zapp/console/src/components/Launch/LaunchForm/ResumeSignalForm.tsx:48]: update import

**Review by olga-union — COMMENTED** [review]
Nice job! Few tiny things left, and don't forget to checkout `package.json`s you accidentally committed 😃

**Comment by james-union** [comment]
@olga-union I've fixed your comments. Just FYI, the flyteidl version is not yet updated because if it points to 1.1.14, it fails to build as `Core.SignalIdentifier` is not defined. 
For the tests, I've fixed `PausedTasksComponent` & `NodeExecutionActions`

Still working on `ResumeSignalForm`

**Review by olga-union — APPROVED** [review]

**Pull request merged**